### PR TITLE
k8s integration tests

### PR DESCRIPTION
**Pull request by juliens**
### What does this PR do?

Add integration tests on kubernetes.

### Motivation

Have tests on kubernetes providers

### More

- [x] Added/updated tests

### Additional Notes

For this, we use https://github.com/rancher/k3s.

As k3s generate a  `kubeconfig.yaml`, I add the ability to use the `KUBECONFIG` env var in Traefik.

### CODE REVIEW

File: integration/fixtures/k8s/test.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            - containerPort: 80
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: whoami
+spec:
+  ports:
+    - name: http
+      targetPort: 80
```
Comment (dtomcej) [inline integration/fixtures/k8s/test.yml]: This line is not needed if `port = targetPort`

File: integration/k8s_test.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+			log.WithoutContext().Debugf("The custom-roles configMap contained K8s object types which are not supported! Skipping object with type: %s", groupVersionKind.Kind)
+		} else {
+			retVal = append(retVal, obj)
+		}
+	}
+	return retVal
+}
+
+func (s *K8sSuite) TestSimpleDefaultConfig(c *check.C) {
+	req := testhelpers.MustNewRequest(http.MethodGet, "https://127.0.0.1:6443", nil)
```
Comment (dtomcej) [inline integration/k8s_test.go]: should `https://127.0.0.1:6443` be a constant?

File: integration/k8s_test.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	err := try.RequestWithTransport(req, time.Second*60, &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}, try.StatusCodeIs(http.StatusUnauthorized))
```
Comment (dtomcej) [inline integration/k8s_test.go]: should ` time.Second*60` be hard-coded?

File: integration/k8s_test.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	c.Assert(err, checker.IsNil)
+
+	abs, err := filepath.Abs("./resources/compose/output/kubeconfig.yaml")
+	c.Assert(err, checker.IsNil)
+
+	err = try.Do(time.Second*60, try.DoCondition(func() error {
```
Comment (dtomcej) [inline integration/k8s_test.go]: should `time.Second*60` be hard coded?

File: integration/k8s_test.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	c.Assert(err, checker.IsNil)
+
+	yamlContent, err := ioutil.ReadFile("./fixtures/k8s/test.yml")
+	c.Assert(err, checker.IsNil)
+
+	k8sObjects := parseK8sYaml(yamlContent)
+	for _, obj := range k8sObjects {
+		switch o := obj.(type) {
+		case *v1beta12.Deployment:
+			_, err := clientset.ExtensionsV1beta1().Deployments("default").Create(o)
```
Comment (dtomcej) [inline integration/k8s_test.go]: should the namespace be hard-coded? `("default")`?

File: provider/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -68,10 +68,13 @@ func (p *Provider) newK8sClient(ctx context.Context, ingressLabelSelector string
...
-		log.FromContext(ctx).Infof("Creating in-cluster Provider client%s", withEndpoint)
+	switch {
+	case os.Getenv("KUBERNETES_SERVICE_HOST") != "" && os.Getenv("KUBERNETES_SERVICE_PORT") != "":
 		cl, err = newInClusterClient(p.Endpoint)
-	} else {
+	case os.Getenv("KUBECONFIG") != "":
+		log.FromContext(ctx).Infof("Creating cluster-external Provider client from KUBECONFIMG %s", os.Getenv("KUBECONFIG"))
+		cl, err = newExternalClusterClientFromFile(os.Getenv("KUBECONFIG"))
+	default:
 		log.FromContext(ctx).Infof("Creating cluster-external Provider client%s", withEndpoint)
```
Comment (dtomcej) [inline provider/kubernetes/kubernetes.go]:
```suggestion
		log.FromContext(ctx).Infof("Creating cluster-external Provider client: %s", withEndpoint)
```

File: integration/k8s_test.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	s.createComposeProject(c, "k8s")
+	s.composeProject.Start(c)
+}
+
+func (s *K8sSuite) TearDownSuite(c *check.C) {
+	s.composeProject.Stop(c)
+	os.Remove("./resources/compose/output/kubeconfig.yaml")
+}
+
+func parseK8sYaml(fileR []byte) []runtime.Object {
```
[the remark this answers is not in the document]
Comment (juliens) [reply integration/k8s_test.go]: I tried to find the best solution and I found this https://github.com/kubernetes/client-go/issues/193

File: provider/kubernetes/kubernetes.go (hunk 6 above)
Comment (juliens) [reply provider/kubernetes/kubernetes.go]: in fact the `withEndpoint` is filled before with 
```golang
withEndpoint := ""
if p.Endpoint != "" {
    withEndpoint = fmt.Sprintf(" with endpoint %v", p.Endpoint)
}
```

File: integration/k8s_test.go (hunk 4 above)
Comment (juliens) [reply integration/k8s_test.go]: I think we nearly never used constant for time in integration tests.
so :thinking:

File: provider/kubernetes/client.go [hunk 8]
```diff
@@ -1,12 +1,12 @@
 package kubernetes
 
 import (
-	"errors"
 	"fmt"
 	"io/ioutil"
 	"time"
 
 	"github.com/containous/traefik/old/log"
+	"github.com/pkg/errors"
```
Comment (ldez) [inline provider/kubernetes/client.go]: not the good package `github.com/pkg/errors ` -> `errors`

File: provider/kubernetes/client.go [hunk 9]
```diff
@@ -78,6 +79,15 @@ func newInClusterClient(endpoint string) (*clientImpl, error) {
 	return createClientFromConfig(config)
 }
 
+func newExternalClusterClientFromFile(file string) (*clientImpl, error) {
+	configFromFlags, err := clientcmd.BuildConfigFromFlags("", file)
+
```
Comment (ldez) [inline provider/kubernetes/client.go]:
```suggestion
```

File: provider/kubernetes/client.go [hunk 10]
```diff
@@ -99,11 +109,11 @@ func newExternalClusterClient(endpoint, token, caFilePath string) (*clientImpl,
 
 		config.TLSClientConfig = rest.TLSClientConfig{CAData: caData}
 	}
-
 	return createClientFromConfig(config)
 }
 
 func createClientFromConfig(c *rest.Config) (*clientImpl, error) {
+
```
Comment (ldez) [inline provider/kubernetes/client.go]:
```suggestion
```

**Comment by emilevauge** [comment]
> Besides k3s, there's also the possibility to use "kind". Mostly FYI though in case we run into troubles with k3s since it's still fairly new.

Indeed, but we were using kind in integration tests on another project and we ended up switching to k3s because:
- kind is a lot heavier
- kind does not provide a simple way to launch it using compose

**Comment by timoreimann** [comment]
@emile that's very interesting to know, thanks. :+1:

**Pull request merged**